File: 20251220 - UI_MAP_BUILDER_INTEGRATION/topologies/stepped_island_clusters.py
# src/map_generator/topologies/stepped_island_clusters.py

import random
from .base_topology import BaseTopology
import copy
from src.map_generator.models.path_info import PathInfo, Coord
from src.utils.geometry import add_vectors, FORWARD_X, UP_Y, FORWARD_Z

class SteppedIslandClustersTopology(BaseTopology):
    """
    Tạo ra nhiều cụm đảo, với mỗi cụm ở một độ cao khác nhau,
    được nối với nhau bằng các bậc thang.
    """

    def _create_island(self, top_left_corner: Coord, size: int = 2) -> list[Coord]:
        """Tạo một hòn đảo hình vuông."""
        x, y, z = top_left_corner
        coords = []
        for i in range(size):
            for j in range(size):
                coords.append((x + i, y, z + j))
        return coords

    def _create_staircase(self, start_point: Coord, direction: Coord, num_steps: int) -> tuple[list[Coord], list[Coord], list[Coord]]:
        """
        [CHUẨN HÓA] Tạo cầu thang rỗng, đi lên theo hướng cho trước.
        Returns:
            - path_coords: Đường đi của người chơi.
            - surface_coords: Các khối bề mặt của bậc thang (để đặt nền).
            - obstacle_coords: Các khối vật cản (chỉ khối trigger cho bước nhảy).
        """
        path = []
        surfaces = []
        obstacles = []
        current_pos = start_point
        for _ in range(num_steps):
            # Bước 1: Đi ngang
            current_pos = add_vectors(current_pos, direction)
            path.append(current_pos)
            obstacles.append(current_pos) # Khối này là vật cản để nhảy LÊN
            # Bước 2: Điểm đáp sau khi nhảy (đi lên)
            current_pos = add_vectors(current_pos, UP_Y)
            path.append(current_pos)
            surfaces.append(current_pos) # Chỉ bề mặt trên cùng là khối vật lý
        return path, surfaces, obstacles

    def _create_bridge(self, start_point: Coord, end_point: Coord) -> list[Coord]:
        """Tạo một cây cầu phẳng từ start đến end theo trục X."""
        path = []
        curr_x, y, z = start_point
        end_x = end_point[0]
        step = 1 if end_x > curr_x else -1
        # Bỏ qua điểm đầu tiên vì nó đã tồn tại
        while curr_x != end_x:
            curr_x += step
            path.append((curr_x, y, z))
        return path

    def generate_path_info_variants(self, grid_size: tuple, params: dict, max_variants: int) -> "Iterator[PathInfo]":
        """
        [KIẾN TRÚC MỚI] Tạo ra các biến thể với số lượng cụm và đảo khác nhau.
        """
        count = 0
        # Lặp qua các cấu hình có thể có
        for num_c in range(2, 4):
            for isl_per_c in range(2, 4):
                if count >= max_variants: return
                
                variant_params = copy.deepcopy(params)
                variant_params['num_clusters'] = num_c
                variant_params['islands_per_cluster'] = isl_per_c
                yield self.generate_path_info(grid_size, variant_params)
                count += 1

    def generate_path_info(self, grid_size: tuple, params: dict) -> PathInfo:
        """Sinh ra các cụm đảo ở các độ cao khác nhau."""
        print("    LOG: Generating 'stepped_island_clusters' topology...")

        num_clusters = params.get('num_clusters', 2)
        islands_per_cluster = params.get('islands_per_cluster', 2)
        cluster_spacing = params.get('cluster_spacing', 8)
        height_step = params.get('height_step', 2)

        start_x = 2
        start_z = grid_size[2] // 2 - (islands_per_cluster * 2) # Căn giữa cụm đảo
        y = 0

        path_coords: list[Coord] = []
        placement_coords: list[Coord] = []
        obstacles = []
        islands: list[list[Coord]] = [] # [FIX] Collect islands for metadata

        # Tạo đảo đầu tiên làm điểm bắt đầu
        first_island = self._create_island((start_x, y, start_z), size=2)
        placement_coords.extend(first_island)
        islands.append(first_island) # [FIX]
        start_pos = (start_x, y, start_z)
        path_coords.append(start_pos)
        last_exit_point = (start_x + 1, y, start_z + 1) # Góc dưới bên phải của đảo đầu tiên
        path_coords.append(last_exit_point)

        for i in range(num_clusters):
            # Tính toán độ cao cho cụm đảo hiện tại
            island_base_y = y + i * height_step
            
            # Nối các đảo trong cụm
            for j in range(islands_per_cluster):
                # Điểm vào của đảo tiếp theo
                next_island_entry = (last_exit_point[0] + 3, island_base_y, last_exit_point[2])
                
                # Tạo cầu nối
                bridge = self._create_bridge(last_exit_point, next_island_entry)
                path_coords.extend(bridge)
                placement_coords.extend(bridge)
                
                # Tạo đảo
                island = self._create_island(next_island_entry, size=2)
                path_coords.append(next_island_entry)
                placement_coords.extend(island)
                islands.append(island) # [FIX]
                
                # Cập nhật điểm ra cho đảo/cầu tiếp theo
                last_exit_point = (next_island_entry[0] + 1, island_base_y, next_island_entry[2] + 1)
                path_coords.append(last_exit_point)

            # Nếu chưa phải cụm cuối, tạo cầu thang đi lên
            if i < num_clusters - 1:
                # [CHUẨN HÓA] Tạo cầu thang và xử lý kết quả
                stair_path, stair_surfaces, stair_obstacles = self._create_staircase(last_exit_point, FORWARD_X, height_step)
                
                path_coords.extend(stair_path)
                # Bề mặt bậc thang cũng là nơi có thể đặt nền móng, nhưng không đặt vật phẩm
                placement_coords.extend(stair_surfaces)
                # Các khối bề mặt được coi là vật cản
                obstacles.extend([{"pos": pos, "is_surface_obstacle": False} for pos in stair_obstacles])
                
                # Cập nhật điểm ra là điểm cuối của cầu thang
                last_exit_point = stair_path[-1] if stair_path else last_exit_point

        target_pos = last_exit_point
        
        # [NEW] Semantic positions for segment_based and island_replication strategies
        # Calculate island centers
        island_centers = []
        for island in islands:
            if island:
                # Center is approximately the first coordinate
                island_centers.append(island[0])
        
        semantic_positions = {
            'start_island': start_pos,
            'end_island': target_pos,
            'optimal_start': 'start_island',
            'optimal_end': 'end_island',
            'valid_pairs': [
                {
                    'name': 'island_hop_easy',
                    'start': 'start_island',
                    'end': 'end_island',
                    'path_type': 'linear_islands',
                    'strategies': ['segment_based', 'island_replication'],
                    'difficulty': 'EASY',
                    'teaching_goal': 'Simple island-to-island traversal'
                },
                {
                    'name': 'cluster_traverse_medium',
                    'start': 'start_island',
                    'end': 'end_island',
                    'path_type': 'cluster_based',
                    'strategies': ['segment_based', 'height_variation'],
                    'difficulty': 'MEDIUM',
                    'teaching_goal': 'Traverse clusters with height awareness'
                },
                {
                    'name': 'all_islands_hard',
                    'start': 'start_island',
                    'end': 'end_island',
                    'path_type': 'full_traversal',
                    'strategies': ['segment_based', 'island_replication'],
                    'difficulty': 'HARD',
                    'teaching_goal': 'Visit all islands with identical patterns'
                }
            ]
        }
        
        # Add island positions
        for i, center in enumerate(island_centers):
            semantic_positions[f'island_{i}'] = center
        
        # [FIX] Populate metadata for IslandPlacerStrategy
        # Convert islands list to segments for segment_based strategy
        segments = [[coord] for island in islands for coord in island]
        
        metadata = {
            "islands": islands,
            "topology_type": "stepped_island_clusters",
            "num_clusters": num_clusters,
            "segments": segments,  # [NEW]
            "semantic_positions": semantic_positions,  # [NEW]
        }

        return PathInfo(
            start_pos=start_pos,
            target_pos=target_pos,
            path_coords=list(dict.fromkeys(path_coords)),
            placement_coords=list(set(placement_coords)),
            obstacles=obstacles,
            metadata=metadata
        )

File: packages/academic-map-generator/src/generator/topologies/SteppedIslandClusters.ts
/**
 * Stepped Island Clusters Topology (PORTED FROM PYTHON)
 * Creates clusters of islands at different heights connected by staircases.
 * Ideal for teaching nested loops and height-based navigation.
 */

import { BaseTopology } from './BaseTopology';
import { IPathInfo, Coord } from '../types';

// Direction vectors
const FORWARD_X: [number, number, number] = [1, 0, 0];
const UP_Y: [number, number, number] = [0, 1, 0];

const addVectors = (a: Coord, b: [number, number, number]): Coord => 
  [a[0] + b[0], a[1] + b[1], a[2] + b[2]];

export class SteppedIslandClustersTopology extends BaseTopology {
  *generatePathInfoVariants(
    gridSize: [number, number, number],
    params: Record<string, any>,
    maxVariants: number
  ): Generator<IPathInfo> {
    let count = 0;
    
    // Loop through possible configurations
    for (let numClusters = 2; numClusters <= 3; numClusters++) {
      for (let islandsPerCluster = 2; islandsPerCluster <= 3; islandsPerCluster++) {
        if (count >= maxVariants) return;
        yield this.generatePathInfo(gridSize, {
          ...params,
          num_clusters: numClusters,
          islands_per_cluster: islandsPerCluster
        });
        count++;
      }
    }
  }

  /**
   * Create a square island at the given top-left corner.
   */
  private _createIsland(topLeftCorner: Coord, size: number = 2): Coord[] {
    const [x, y, z] = topLeftCorner;
    const coords: Coord[] = [];
    for (let i = 0; i < size; i++) {
      for (let j = 0; j < size; j++) {
        coords.push([x + i, y, z + j]);
      }
    }
    return coords;
  }

  /**
   * Create a staircase going up in the given direction.
   * Returns path_coords, surface_coords (for placement), and obstacle_coords.
   */
  private _createStaircase(
    startPoint: Coord, 
    direction: [number, number, number], 
    numSteps: number
  ): {
    pathCoords: Coord[];
    surfaceCoords: Coord[];
    obstacleCoords: Coord[];
  } {
    const pathCoords: Coord[] = [];
    const surfaceCoords: Coord[] = [];
    const obstacleCoords: Coord[] = [];
    let currentPos: Coord = [...startPoint];

    for (let step = 0; step < numSteps; step++) {
      // Step 1: Move horizontally (this becomes the obstacle to jump from)
      currentPos = addVectors(currentPos, direction);
      pathCoords.push([...currentPos]);
      obstacleCoords.push([...currentPos]);

      // Step 2: Landing position after jump (move up)
      currentPos = addVectors(currentPos, UP_Y);
      pathCoords.push([...currentPos]);
      surfaceCoords.push([...currentPos]); // Only the top surface is physical
    }

    return { pathCoords, surfaceCoords, obstacleCoords };
  }

  /**
   * Create a flat bridge from start to end along X axis.
   */
  private _createBridge(startPoint: Coord, endPoint: Coord): Coord[] {
    const path: Coord[] = [];
    let currX = startPoint[0];
    const y = startPoint[1];
    const z = startPoint[2];
    const endX = endPoint[0];
    const step = endX > currX ? 1 : -1;

    // Skip first point as it already exists
    while (currX !== endX) {
      currX += step;
      path.push([currX, y, z]);
    }
    return path;
  }

  generatePathInfo(gridSize: [number, number, number], params: Record<string, any>): IPathInfo {
    console.log("    LOG: Generating 'stepped_island_clusters' topology...");

    const numClusters = params.num_clusters || 2;
    const islandsPerCluster = params.islands_per_cluster || 2;
    const clusterSpacing = params.cluster_spacing || 8;
    const heightStep = params.height_step || 2;

    const startX = params.start_x || 2;
    const startZ = params.start_z || Math.floor(gridSize[2] / 2) - islandsPerCluster;
    let y = 0;

    const pathCoords: Coord[] = [];
    const placementCoords: Coord[] = [];
    const obstacles: any[] = [];
    const islandsList: Coord[][] = [];

    // Create first island as starting point
    const firstIsland = this._createIsland([startX, y, startZ], 2);
    placementCoords.push(...firstIsland);
    islandsList.push(firstIsland);
    
    const startPos: Coord = [startX, y, startZ];
    pathCoords.push(startPos);
    
    let lastExitPoint: Coord = [startX + 1, y, startZ + 1]; // Bottom-right corner of first island
    pathCoords.push(lastExitPoint);

    for (let cluster = 0; cluster < numClusters; cluster++) {
      // Calculate base Y for this cluster
      const islandBaseY = y + cluster * heightStep;

      for (let island = 0; island < islandsPerCluster; island++) {
        // Entry point of next island
        const nextIslandEntry: Coord = [lastExitPoint[0] + 3, islandBaseY, lastExitPoint[2]];

        // Create bridge to next island
        const bridge = this._createBridge(lastExitPoint, nextIslandEntry);
        pathCoords.push(...bridge);
        placementCoords.push(...bridge);

        // Create island
        const islandCoords = this._createIsland(nextIslandEntry, 2);
        pathCoords.push(nextIslandEntry);
        placementCoords.push(...islandCoords);
        islandsList.push(islandCoords);

        // Update exit point for next island/bridge
        lastExitPoint = [nextIslandEntry[0] + 1, islandBaseY, nextIslandEntry[2] + 1];
        pathCoords.push(lastExitPoint);
      }

      // Create staircase to next cluster (if not last cluster)
      if (cluster < numClusters - 1) {
        const { pathCoords: stairPath, surfaceCoords, obstacleCoords } = 
          this._createStaircase(lastExitPoint, FORWARD_X, heightStep);

        pathCoords.push(...stairPath);
        placementCoords.push(...surfaceCoords);
        
        // Add obstacles with proper structure
        for (const pos of obstacleCoords) {
          obstacles.push({ pos, is_surface_obstacle: false });
        }

        // Update exit point to end of staircase
        lastExitPoint = stairPath.length > 0 ? stairPath[stairPath.length - 1] : lastExitPoint;
      }
    }

    const targetPos = lastExitPoint;

    // Calculate island centers for semantic positions
    const islandCenters: Coord[] = islandsList.map(island => island[0]);

    // Semantic positions (matching Python)
    const semantic_positions: Record<string, any> = {
      start_island: startPos,
      end_island: targetPos,
      optimal_start: 'start_island',
      optimal_end: 'end_island',
      valid_pairs: [
        {
          name: 'island_hop_easy',
          start: 'start_island',
          end: 'end_island',
          path_type: 'linear_islands',
          strategies: ['segment_based', 'island_replication'],
          difficulty: 'EASY',
          teaching_goal: 'Simple island-to-island traversal'
        },
        {
          name: 'cluster_traverse_medium',
          start: 'start_island',
          end: 'end_island',
          path_type: 'cluster_based',
          strategies: ['segment_based', 'height_variation'],
          difficulty: 'MEDIUM',
          teaching_goal: 'Traverse clusters with height awareness'
        },
        {
          name: 'all_islands_hard',
          start: 'start_island',
          end: 'end_island',
          path_type: 'full_traversal',
          strategies: ['segment_based', 'island_replication'],
          difficulty: 'HARD',
          teaching_goal: 'Visit all islands with identical patterns'
        }
      ]
    };

    // Add island positions dynamically
    for (let i = 0; i < islandCenters.length; i++) {
      semantic_positions[`island_${i}`] = islandCenters[i];
    }

    // Deduplicate placement coords
    const uniquePlacement = this._deduplicateCoords(placementCoords);
    
    // Use the manually built pathCoords which includes all valid steps and jumps
    // BFS fails here because staircase "jumps" create gaps in the walkable grid
    const uniquePath = this._deduplicateCoords(pathCoords);

    // Use original Y coordinates as requested (aligning with block surface)
    // The visualization engine handles the rendering offset if needed
    const finalPath: Coord[] = uniquePath;
    const finalStart: Coord = startPos;
    const finalTarget: Coord = targetPos;

    // Segment analysis
    const segments = islandsList.map(island => [...island]);
    const lengths = segments.map(s => s.length);
    const segment_analysis = {
      num_segments: segments.length,
      lengths,
      types: segments.map(() => 'island'),
      min_length: lengths.length > 0 ? Math.min(...lengths) : 0,
      max_length: lengths.length > 0 ? Math.max(...lengths) : 0,
      avg_length: lengths.length > 0 ? lengths.reduce((a, b) => a + b, 0) / lengths.length : 0
    };

    return {
      start_pos: finalStart,
      target_pos: finalTarget,
      path_coords: finalPath,        // DYNAMIC: manually constructed path including jumps
      placement_coords: uniquePlacement, // STATIC: all walkable tiles
      obstacles: obstacles,
      metadata: {
        topology_type: 'stepped_island_clusters',
        islands: islandsList,
        num_clusters: numClusters,
        islands_per_cluster: islandsPerCluster,
        height_step: heightStep,
        cluster_spacing: clusterSpacing,
        segments: segments,
        segment_analysis,
        semantic_positions
      }
    };
  }
}
